client/src/main/java/org/eclipse/hono/client/impl/DestinationCommandConsumer.java

```diff
@@ -0,0 +1,318 @@
+     * command consumer.
+     * 
+     * @param gatewayManagedDeviceId The id of a device connected to a gateway. May be {@code null}.
+     *                               If a handler for that device exists, it will be returned here.
+     *                               Otherwise the handler for <em>this</em> command consumers
+     *                               device id will be returned (if set).
+     * @return The handler or {@code null}.
+     */
+    public CommandHandlerWrapper getCommandHandlerOrDefault(final String gatewayManagedDeviceId) {
+        if (gatewayManagedDeviceId != null && !gatewayManagedDeviceId.equals(gatewayOrDeviceId)) {
+            final CommandHandlerWrapper handler = commandHandlers.get(gatewayManagedDeviceId);
+            if (handler != null) {
+                LOG.trace("using device specific command handler [gateway-id: {}, device-id: {}]", gatewayOrDeviceId,
+                        gatewayManagedDeviceId);
+                return handler;
+            }
+        }
+        return commandHandlers.get(gatewayOrDeviceId);
+    }
+
+    /**
+     * Gets the contained command handlers.
+     *
+     * @return The command handlers.
+     */
+    public Collection<CommandHandlerWrapper> getCommandHandlers() {
+        return commandHandlers.values();
+    }
+
+    private void onRemoteClose(final Handler<String> remoteCloseHandler, final String event) {
+        remoteCloseHandler.handle(event);
+        commandHandlers.values().forEach(handler -> {
+            handler.getRemoteCloseHandler().handle(null);
+        });
+    }
+
+    /**
+     * Removes the handler for the given device id and closes this consumer if there are
+     * no remaining handlers left.
+     *
+     * @param deviceId The device id of the handler to remove.
+     * @param resultHandler A handler that is called back with the result of the operation.
+     */
+    public void removeHandlerAndCloseConsumerIfEmpty(final String deviceId, final Handler<AsyncResult<Void>> resultHandler) {
+        final CommandHandlerWrapper removedHandler = commandHandlers.remove(deviceId);
+        LOG.trace("Removed handler for device {} on consumer {}: {}", deviceId, this.gatewayOrDeviceId, removedHandler != null);
+        if (removedHandler != null && commandHandlers.isEmpty()) {
+            LOG.trace("all command handlers removed for gateway consumer, closing link [gateway-id: {}, device-id of removed handler: {}]",
+                    this.gatewayOrDeviceId, deviceId);
+            closedCalled.set(true);
+            close(resultHandler);
+        } else if (resultHandler != null) {
+            resultHandler.handle(Future.succeededFuture());
+        }
+    }
+
+    /**
+     * Creates a new command consumer.
+     * <p>
+     * The handler to be invoked by the created consumer will have to be subsequently added
+     * via the {@link #addDeviceSpecificCommandHandler(String, String, Handler, Handler)} method.
```

Comment:
For the use case of the `CommandConsumerFactoryImpl` (ie. the *only* use case for now), the current method is more convenient. I have to potentially add the handler outside of the `create` method anyway because it could be that I already have a cached consumer instance.

Reply:
I see